src/context.jl

```diff
@@ -0,0 +1,169 @@
+using IRTools: isexpr, IR, @dynamo, postwalk
+using IRTools: meta, Pipe, finish, Variable, self
+using MacroTools: @forward
+
+import Base.Broadcast.broadcasted
+import Base.Broadcast.materialize
+import Base.Broadcast.Broadcasted
+
+# TODO use a WeakKeyDict
+struct CUDACtx
+  array_bank::IdDict{Array,CuArray}
+end
+
+CUDACtx() = CUDACtx(IdDict{Array,CuArray}())
+
+# Display fns for debugging, remove before committing
+function Base.summary(io::IO, c::CUDACtx)
+  print(io, "IR Context for CUDA ")
+  summary(io, c.array_bank)
+end
+
+function Base.show(io::IO, c::CUDACtx)
+  print(io, "IR Context for CUDA ")
+  display(c.array_bank)
+end
+
+@forward CUDACtx.array_bank Base.getindex, Base.iterate,
+			Base.setindex!, Base.empty!,
+			Base.length,
+			Base.first, Base.last, Base.haskey
+
+function _resize!(a::Array, sz::NTuple{<:Any,Integer})
+  ccall(:jl_array_grow_end, Cvoid, (Any, UInt), a, prod(sz))
+  ptr = convert(Ptr{Csize_t},pointer_from_objref(a))
+  for i = 1:length(sz)
+    unsafe_store!(ptr+8*(i+2), sz[i])
+  end
+  return a
+end
+
+function refill!(a::Array, b::CuArray)
+  _resize!(a, size(b))
+  copy!(a, b)
+end
+
+function cache(cx, x::CuArray{T,N})::Array{T,N} where {T,N}
+  cpu = Array{T,N}(undef, ntuple(_->0,N))
+  cx[cpu] = x
+  return cpu
+end
+cache(cx, f) = f
+
+for f in (:+, :-, :*, :/)
+  @eval function (c::CUDACtx)(::typeof($f), a::AbstractArray, b::AbstractArray)
+    ga = get_cached(array_bank, a)
+    gb = get_cached(array_bank, b)
+    cache(array_bank, $f(ga, gb))
+  end
+end
+
+function get_cached(array_bank, arr::Array{T,N})::CuArray{T,N} where {T,N}
+  haskey(array_bank, arr) ?
+    array_bank[arr] :
+    (array_bank[arr] = CuArray(arr))
+end
+
+function (c::CUDACtx)(::typeof(broadcasted), f, args...)
+  gargs = map(x -> get_cached(array_bank, x), args)
+  broadcasted(f, gargs...) |> x -> cache(array_bank, x)
+end
+
+function (c::CUDACtx)(::typeof(getproperty), o, s::Symbol)
```

Comment:
We shouldn't need this function, along with `getfield` etc; arrays should only converted when they are used as inputs to CUDA primitives. This would also return a `CuArray` to the users program.

Reply:
Done